### ISSUE TITLE
CustomPaint: Shader with SweepGradient crashes on WEB, but runs OK on device and desktop

### ISSUE THREAD

**Issue by rydmike** [flutter/flutter#57752]
## Steps to Reproduce

1. Copy the example code from this gist:
https://gist.github.com/rydmike/dc8a915f169455463e4e2016619e00bb

2. Build it for WEB. When the **"Crash me on WEB"** switch is turned on, the application will crash if it is running on Web, but it will run OK and produce the expected result on a device or desktop.

You can try it it DartPad here;
https://dartpad.dartlang.org/dc8a915f169455463e4e2016619e00bb

![image](https://user-images.githubusercontent.com/39990307/82606220-63952900-9bbf-11ea-8dda-ec0833bc1a42.png)

But not see the crash reason. The crash occurs but when using Flutter WEB DomCanvas and CanvasKit.

**TLDR**
The second method shown in the example of producing the color circle is also in use by a some packages on pub.dev. As shown further below the merits of using this method can be argued against. Regardless of that, Flutter code that uses rather basic API calls, runs and produces expected output on devices and desktops, should probably not crash on Web.

The line that causes the crash on Web is isolated and highlighted below:
![image](https://user-images.githubusercontent.com/39990307/82603065-adc7db80-9bba-11ea-89af-76111ef2dfac.png)

------

<details>
  <summary>Logs and version</summary>

**Error log from web crash below**
```
[   +1 ms] The following JSNoSuchMethodError was thrown during paint():
[  +69 ms] NoSuchMethodError: invalid member on null: '_get'
[  +46 ms] The relevant error-causing widget was:
[   +2 ms]   CustomPaint file:///C:/Users/mryds/OneDrive/code/flutter/_mr/scroll_pan/lib/main.dart:73:24
[  +77 ms] When the exception was thrown, this was the stack:
[   +1 ms] C:/b/s/w/ir/cache/builder/src/out/host_debug/flutter_web_sdk/lib/_engine/engine.dart 239:26         toMatrix32
[   +1 ms] C:/b/s/w/ir/cache/builder/src/out/host_debug/flutter_web_sdk/lib/ui/src/ui/painting.dart 1186:71    sweep
[        ] packages/flutter/src/painting/gradient.dart 920:24                                                  createShader
[        ] packages/scroll_pan/main.dart 180:29                                                                paint
[        ] packages/flutter/src/rendering/custom_paint.dart 531:12                                             [_paintWithPainter]
[  +36 ms] packages/flutter/src/rendering/custom_paint.dart 572:7                                              paint
[  +38 ms] packages/flutter/src/rendering/object.dart 2264:7                                                   [_paintWithContext]
[   +3 ms] packages/flutter/src/rendering/object.dart 184:12                                                   paintChild
[   +1 ms] packages/flutter/src/rendering/proxy_box.dart 131:14                                                paint
[   +1 ms] packages/flutter/src/rendering/object.dart 2264:7                                                   [_paintWithContext]
[   +1 ms] packages/flutter/src/rendering/object.dart 184:12                                                   paintChild
[        ] packages/flutter/src/rendering/shifted_box.dart 70:14                                               paint
[   +1 ms] packages/flutter/src/rendering/object.dart 2264:7                                                   [_paintWithContext]
[ +101 ms] packages/flutter/src/rendering/object.dart 184:12                                                   paintChild
[   +4 ms] packages/flutter/src/rendering/box.dart 2515:14                                                     defaultPaint
[   +1 ms] packages/flutter/src/rendering/flex.dart 950:7                                                      paint
[   +1 ms] packages/flutter/src/rendering/object.dart 2264:7                                                   [_paintWithContext]
[   +1 ms] packages/flutter/src/rendering/object.dart 184:12                                                   paintChild
[        ] packages/flutter/src/widgets/single_child_scroll_view.dart 548:16                                   paintContents
[        ] packages/flutter/src/rendering/object.dart 391:12                                                   pushLayer
[        ] packages/flutter/src/rendering/object.dart 447:7                                                    pushClipRect
[        ] packages/flutter/src/widgets/single_child_scroll_view.dart 552:16                                   paint
[        ] packages/flutter/src/rendering/object.dart 2264:7                                                   [_paintWithContext]
[        ] packages/flutter/src/rendering/object.dart 135:10                                                   _repaintCompositedChild
[   +1 ms] packages/flutter/src/rendering/object.dart 95:5                                                     repaintCompositedChild
[  +71 ms] packages/flutter/src/rendering/object.dart 980:29                                                   flushPaint
[   +2 ms] packages/flutter/src/rendering/binding.dart 404:19                                                  drawFrame
[   +1 ms] packages/flutter/src/widgets/binding.dart 865:13                                                    drawFrame
[   +1 ms] packages/flutter/src/rendering/binding.dart 284:5                                                   [_handlePersistentFrameCallback]
[   +1 ms] packages/flutter/src/scheduler/binding.dart 1113:15                                                 [_invokeFrameCallback]
[   +1 ms] packages/flutter/src/scheduler/binding.dart 1052:9                                                  handleDrawFrame
[        ] packages/flutter/src/scheduler/binding.dart 968:5                                                   [_handleDrawFrame]
[        ] C:/b/s/w/ir/cache/builder/src/out/host_debug/flutter_web_sdk/lib/_engine/engine/window.dart 715:13  _invoke
[        ] C:/b/s/w/ir/cache/builder/src/out/host_debug/flutter_web_sdk/lib/_engine/engine/window.dart 338:5   invokeOnDrawFrame
[        ] C:/b/s/w/ir/cache/builder/src/out/host_debug/flutter_web_sdk/lib/_engine/engine.dart 211:18         <fn>
[  +72 ms] The following RenderObject was being processed when the exception was fired: RenderCustomPaint#891a8:
[   +3 ms]   creator: CustomPaint ← SizedBox ← Center ← Column ← _SingleChildViewport ←
[   +2 ms]     IgnorePointer-[GlobalKey#eb731] ← Semantics ← _PointerListener ← Listener ← _GestureSemantics ←
[   +1 ms]     RawGestureDetector-[LabeledGlobalKey<RawGestureDetectorState>#84767] ← _PointerListener ← ⋯
[        ]   parentData: <none> (can use size)
[        ]   constraints: BoxConstraints(w=170.0, h=170.0)
[        ]   size: Size(170.0, 170.0)
[        ] This RenderObject has no descendants.
[        ] ════════════════════════════════════════════════════════════════════════════════════════════════════
[   +1 ms] Another exception was thrown: Assertion failed: org-dartlang-sdk:///flutter_web_sdk/lib/_engine/engine/bitmap_canvas.dart:788:12
[  +54 ms] Another exception was thrown: NoSuchMethodError: invalid member on null: '_get'
[   +2 ms] Another exception was thrown: PersistedPhysicalShape: is in an unexpected state.
[+1672 ms] Another exception was thrown: PersistedScene: is in an unexpected state.
[  +11 ms] Another exception was thrown: PersistedScene: is in an unexpected state.
[  +23 ms] Another exception was thrown: PersistedScene: is in an unexpected state.
[  +71 ms] Another exception was thrown: PersistedScene: is in an unexpected state.

```

**Used version**
The above report was made on latest master channel shown below. However, as the DartPad example shows, it also crashes on the version it is running on.

```
flutter doctor -v
[√] Flutter (Channel master, 1.19.0-2.0.pre.133, on Microsoft Windows [Version 10.0.18363.778], locale en-US)
    • Flutter version 1.19.0-2.0.pre.133 at c:\src\flutter
    • Framework revision e92afc16b6 (24 hours ago), 2020-05-21 05:29:01 +0900
    • Engine revision 2d4e83921d
    • Dart version 2.9.0 (build 2.9.0-9.0.dev 40f7a11d89)


[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at C:\Users\mryds\AppData\Local\Android\sdk
    • Platform android-28, build-tools 28.0.3
    • Java binary at: C:\Program Files\Android\Android Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)
    • All Android licenses accepted.

[√] Chrome - develop for the web
    • Chrome at C:\Program Files (x86)\Google\Chrome\Application\chrome.exe

[√] Visual Studio - develop for Windows (Visual Studio Community 2019 16.5.4)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Visual Studio Community 2019 version 16.5.30011.22

[√] Android Studio (version 3.6)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin version 45.1.1
    • Dart plugin version 192.8052
    • Java version OpenJDK Runtime Environment (build 1.8.0_212-release-1586-b04)

[√] IntelliJ IDEA Community Edition (version 2019.2)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2018.3.1
    • Flutter plugin version 35.3.3
    • Dart plugin version 192.7402

[√] VS Code (version 1.45.1)
    • VS Code at C:\Users\mryds\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.10.2

[√] Connected device (4 available)
    • Nexus 7    • 0a99f5e8   • android-arm    • Android 6.0.1 (API 23)
    • Windows    • Windows    • windows-x64    • Microsoft Windows [Version 10.0.18363.778]
    • Web Server • web-server • web-javascript • Flutter Tools
    • Chrome     • chrome     • web-javascript • Google Chrome 81.0.4044.138

• No issues found!
```

</details>

-----

# More info

The Gist example also show an alternative and better way of producing the sought output that does not crash on Web.

The first color circle in the example does not crash on Web:
![image](https://user-images.githubusercontent.com/39990307/82604014-4743bd00-9bbc-11ea-8b18-5e972d26114e.png)

But the second one will crash on Web when the switch **"Crash me on Web"** is turned on. When running on desktop or a device, it will look very similar to the first one:

![image](https://user-images.githubusercontent.com/39990307/82604149-82de8700-9bbc-11ea-951a-0a128b21e8a8.png)

It is however not exactly the same, the second version version using the gradient and shader that crashes on Web, is actually not an entirely correct representation of the color circle. This becomes obvious when the width is of the circle is increased.

![image](https://user-images.githubusercontent.com/39990307/82604378-e799e180-9bbc-11ea-83f5-cf5cf7e25ce7.png)

I have noticed that this SweepGradient + Shader way of producing the color circle can be found in a few packages on pub.dev. While the merits of using this method for wider color circles can be argued, the same code that runs fine without any issues on a device and desktop, using very basic Flutter APIs should probably not crash on Flutter WEB.

**Comment by yjbanov** [comment]
This is a duplicate of https://github.com/flutter/flutter/issues/41389

**Comment by rydmike** [comment]
OK thanks @yjbanov for letting me know and finding the duplicate. I will be tracking its progress instead. Sorry that I did not find it myself before making this report, due to the amount of issues and not being familiar with the back-log, it can sometimes be difficult to locate an already existing issue when reporting a new one. Glad that you caught it.

Also for my particular use case, I found another way of doing what I needed that works on Web. I have however seen that some packages on pub.dev have used to approach shown here. I'll let the packages authors know about this issue as well.

**Comment by yjbanov** [comment]
@rydmike Don't let it stop you from filing issues! It's better to have a duplicate than not having an issue filed at all :)

**Comment by rydmike** [comment]
No worries @yjbanov I will :grinning: 

This particular example case actually crashes on both DomCanvas and CanvasKit (SKIA).
I just learned that in some cases some CustomPaint APIs work in CavnasKit but not on DomCanvas.
It is honestly a bit confusing as to what works where, would be great if there was a status list of what is implemented and supported on each render engine. Is there?

The example in this issue report actually also contain a solution (alternative implementation) for the particular color picker use case. I will therefore refer others that have published packages on pub.dev that use the approach that does not work, to this case as a solution example, if they want to make a version that also works on Flutter web, both DomCanvas and CanvasKit.

**Comment by rydmike** [comment]
Just in case some of the people that I am directing to this issue for how to make color wheel that works in Flutter, I just want to add that I just updated the GIST code for the example that works. 

The new updated version is much cleaner and simpler. The previous version used HSV/HSL color handling code that was borrowed/extracted from dart package TinyColor. Using TinyColor is however not needed for this use case, as the needed HSV/HSL color handling classes already exist in the Flutter SDK, the updated example takes advantage of that and is much simpler and shorter.

**Comment by rydmike** [comment]
I will add references to existing Flutter Color pickers that I have come across, that at time of being added to this list, suffers from the issue described here, when used on Web. The packages' issue can be solved by changed the implementation to the workaround shown in this issue in the DratPad example:

- **Package:** https://pub.dev/packages/flutter_circle_color_picker
- Has open issue on the topic: https://github.com/itome/flutter_circle_color_picker/issues/8

---
...more later
